reconcile/change_owners.py

```diff
@@ -705,36 +757,220 @@ def run(
+            )
+            labels = manage_conditional_label(
+                labels=labels,
+                condition=self_servicable and approved,
+                true_label=APPROVED,
+            )
+            labels = manage_conditional_label(
+                labels=labels,
+                condition=self_servicable and not approved,
+                true_label=AWAITING_APPROVAL,
+            )
+        gl.set_labels_on_merge_request(gitlab_merge_request_id, labels)
+
+    except BaseException:
+        logging.error(traceback.format_exc())
+
+
+def manage_conditional_label(
+    labels: list[str],
+    condition: bool,
+    true_label: Optional[str] = None,
+    false_label: Optional[str] = None,
+) -> list[str]:
+    new_labels = labels.copy()
+    if condition:
+        if true_label and true_label not in labels:
+            new_labels.append(true_label)
+        if false_label and false_label in labels:
+            new_labels.remove(false_label)
+    else:
+        if true_label and true_label in labels:
+            new_labels.remove(true_label)
+        if false_label and false_label not in labels:
+            new_labels.append(false_label)
+    return new_labels
+
+
+@dataclass
+class Decision:
+
+    approve: bool = False
+    hold: bool = False
+
+
+@dataclass
+class ChangeDecision:
+
+    file: FileRef
+    diff: Diff
+    decision: Decision
+
+
+class DecisionCommand(Enum):
+    APPROVED = "/lgtm"
+    CANCEL_APPROVED = "/lgtm cancel"
+    HOLD = "/hold"
+    CANCEL_HOLD = "/hold cancel"
+
+
+def get_approver_decisions(comments: list[dict[str, str]]) -> dict[str, Decision]:
+    decisions_by_users: dict[str, Decision] = defaultdict(Decision)
+    for c in sorted(comments, key=lambda k: k["created_at"]):
+        commenter = c["username"]
+        for line in c.get("body", "").split("\n"):
+            if line == DecisionCommand.APPROVED.value:
+                decisions_by_users[commenter].approve = True
+            if line == DecisionCommand.CANCEL_APPROVED.value:
+                decisions_by_users[commenter].approve = False
+            if line == DecisionCommand.HOLD.value:
+                decisions_by_users[commenter].hold = True
+            if line == DecisionCommand.CANCEL_HOLD.value:
+                decisions_by_users[commenter].hold = False
+    return decisions_by_users
+
+
+def decide_on_changes(
+    changes: list[BundleFileChange], approver_decisions: dict[str, Decision]
+) -> list[ChangeDecision]:
+    diff_decisions = []
+    for c in changes:
+        for d in c.diffs:
+            dc = ChangeDecision(file=c.fileref, diff=d, decision=Decision())
+            diff_decisions.append(dc)
+            for change_type_context in d.covered_by:
+                for approver in change_type_context.approvers:
+                    if approver_decisions[approver.org_username].approve:
+                        dc.decision.approve |= True
+                    if approver_decisions[approver.org_username].hold:
```

Comment:
KeyError possible? Is `approver.org_username` always in `approver_decisions`?

Reply:
the dict is a `defaultdict(Decision)`

Reply:
but you are right. within this context this might be very well a normal dict and KeyErrors might occure. it is somehow hidden that it is a defaultdict. for future safety i will rework this

Reply:
[ad45422](https://github.com/app-sre/qontract-reconcile/pull/2802/commits/ad4542237036e32e505231c7099d9af42ca95634)